### DISCUSSION TITLE
Starting Quest from Script

### DISCUSSION THREAD

**Discussion by ruedi99ms** [spatialsys/spatial-unity-sdk#31] [Help] [answered]
How can I start a SpatialQuest from a C#-Script? I am coding a series of Quests the user can select before starting the game. Therefore, I need to start the quests from code.

Thanks :-)

**Comment by AlanSpatial** [comment]
If your quest is a component you can reference it and start it something like:

```csharp
public SpatialQuest myQuest;

private void StartQuest()
{
    SpatialBridge.questService.quests[myQuest.id].Start();
}
```

See [quest service](https://toolkit.spatial.io/reference/SpatialSys.UnitySDK.IQuestService) for creating or reading quests in c#.

See [IQuest](https://toolkit.spatial.io/reference/SpatialSys.UnitySDK.IQuest) for modifiying a quest with c#.

**Reply by ruedi99ms** [reply]
When I do it like this, I get a KeyNotFoundException:

![grafik](https://github.com/spatialsys/spatial-unity-sdk/assets/105639819/13cbbe40-5ac1-4281-8a7f-8b357222b58d)

![grafik](https://github.com/spatialsys/spatial-unity-sdk/assets/105639819/e6e3b0d8-5e5d-4f01-80fa-6482f54a3a6e)

![grafik](https://github.com/spatialsys/spatial-unity-sdk/assets/105639819/8f5fa6c9-0fe0-4c23-9f8a-510040722a4b)

`public class QuestController : MonoBehaviour
{
    int[] indexSelectedQuests;

    SpatialQuest[] allQuests;
    SpatialQuest[] selectedQuests;

    public SpatialQuest questSDG1;
    public SpatialQuest questSDG2;
    public SpatialQuest questSDG3;
    public SpatialQuest questSDG4;
    public SpatialQuest questSDG5;
    public SpatialQuest questSDG6;
    public SpatialQuest questSDG7;
    public SpatialQuest questSDG8;
    public SpatialQuest questSDG9;
    public SpatialQuest questSDG10;
    public SpatialQuest questSDG11;
    public SpatialQuest questSDG12;
    public SpatialQuest questSDG13;
    public SpatialQuest questSDG14;
    public SpatialQuest questSDG15;
    public SpatialQuest questSDG16;
    public SpatialQuest questSDG17;

    public GameObject cube_prefab;

    private SpatialQuest activatedQuest;


    private int currentQuestNo = 0;

    // Start is called before the first frame update
    void Start()
    {
        indexSelectedQuests = new int[3] {2,1,3}; //Includes the number of the SDG (not the list index!)

        allQuests = new SpatialQuest[17] { questSDG1, questSDG2, questSDG3, questSDG4, questSDG5, questSDG6, questSDG7, questSDG8, questSDG9, questSDG10, questSDG11, questSDG12, questSDG13, questSDG14, questSDG15, questSDG16, questSDG17 };
        selectedQuests = new SpatialQuest[indexSelectedQuests.Length];

        for (int i = 0; i < indexSelectedQuests.Length; i++)
        {
            selectedQuests[i] = allQuests[indexSelectedQuests[i]-1];
        }

        NextQuest();
    }

    // Update is called once per frame
    void Update()
    {
    
    }

    public void NextQuest()
    {
        GameObject tmp = Instantiate(cube_prefab, new Vector3(currentQuestNo, currentQuestNo, 0), Quaternion.identity);

        activatedQuest = selectedQuests[currentQuestNo];
        activatedQuest.onCompletedEvent += NextQuest;

        SpatialBridge.questService.quests[(uint)activatedQuest.id].Start();

        //Start next Quest currently not working

        currentQuestNo++;            
    }`

Any idea why?

**Reply by AlanSpatial** [reply, marked as answer]
oh my bad. I think quests are still initializing on the first frame (during `start()`)

You can wait one frame by replacing your `void Start()` with something like this:
```
IEnumerator Start()
{
    // Wait for next frame
    yield return null;
    
    // do your normal script stuff here
    ...
}
```

**Reply by ruedi99ms** [reply]
That seem to have worked :-) I have updated and cleaned the code

But now I get a new error:

![grafik](https://github.com/spatialsys/spatial-unity-sdk/assets/105639819/0ffd4138-5172-4838-9a01-cb09d27fcbc5)


This happens once the first quest has been completed.

`// Start is called before the first frame update
IEnumerator Start()
{
    yield return null;

    indexSelectedQuests = new int[3] {2,1,3}; //Includes the number of the SDG (not the list index!)

    allQuests = new SpatialQuest[17] { questSDG1, questSDG2, questSDG3, questSDG4, questSDG5, questSDG6, questSDG7, questSDG8, questSDG9, questSDG10, questSDG11, questSDG12, questSDG13, questSDG14, questSDG15, questSDG16, questSDG17 };
    selectedQuests = new SpatialQuest[indexSelectedQuests.Length];

    for (int i = 0; i < indexSelectedQuests.Length; i++)
    {
        selectedQuests[i] = allQuests[indexSelectedQuests[i]-1];
        selectedQuests[i].onCompletedEvent += NextQuest;
    }

    NextQuest();
}

// Update is called once per frame
void Update()
{

}

public void NextQuest()
{
    SpatialBridge.questService.quests[selectedQuests[currentQuestNo].id].Start();

    currentQuestNo++;            
}`

**Reply by ruedi99ms** [reply]
Did found a workaround:

Not starting the next quest in the OnCompleted Event, but checking if the quest is completed in the Update-Function and then starting the next one.
